File: algorithms/Sorting_Algorithms/C++/Merge_sort.cpp
#include<iostream>
using namespace std;

void Merge(int left[],int nl, int right[],int nr , int A[],int n)
{
	 int size_of_left=nl;
	 int size_of_right=nr;
	 int size_of_main=n;
	 
	 int i=0, j=0,k=0;
	 
	 while((i<size_of_left) && (j<size_of_right))
	 {cout<<"in while";
	 	if(left[i]<right[j])
	 	{
	 		A[k]=left[i];
	 		i=i+1;
		 }
		 
		 	else  
	 	{
	 		A[k]=right[j];
	 		j=j+1;
	 		
		 }
		 
		 k=k+1;
		 
	 
	 }
	 //if any indexes are left
	 if(i<size_of_left)
	 {
	 	A[k]=left[i];
	 	i=i+1;
	 	k=k+1;
	 }
	 else if(j<size_of_right)
	 {
	 	A[k]=right[i];
	 	j=j+1;
	 	k=k+1;
	 }
return;	 
}

void Merge_sort(int A[],int siz)
{
	int length= siz;
	if(length<2)
	{
		return;
	}
	int mid=length/2;
	int left[10];
	int right[10];
	
	int nL=mid;
	int nR=mid-length;
	for(int i=0; i<mid; i++)
	{
		left[i]=A[i];
	 
	}
	int x=mid-length;
	int M= mid;
	for(int i=0; i<(x ); i++)
	{
		right[i]=A[M];
		M++;
	 
	}
	Merge_sort(left,nL);
	Merge_sort(right,nR);
	Merge(left,nL, right,nR, A,length);
	
}
void Print(int A[],int n)
{
	cout<<"\n After sort";
	
		for(int i=0; i<n ; i++)
	{
	cout<<"\t"<<A[i];
		
	}
}
int main()

{
	int size;
	cout<<"Enter size of arrray: ";
	cin>>size;
	int A[10];
	cout<<"Enter the values of aray\n";
	for(int i=0; i<size ; i++)
	{
		cout<<"value "<<i<<" : ";
		cin>>A[i];
	}
	Merge_sort(A,size);
	Print(A,size);
	return 0;
}

File: algorithms/ar-msrt/python2/mergesort.py
def merge(arr,lb,ub,m):
    #length of left and right subarrays
    n1,n2 = m - lb + 1,ub - m
    #temp arrays storing left and right partions respectively
    L,R = arr[lb : lb+n1],arr[m+1 : m+1+n2]
    
    #merging temp arrays back to arr
    i,j = 0,0 #initial indices of left  and right subarrays respectively
    k = lb # intial index of merged array 
    while(i<n1 and j<n2):
        if L[i] <= R[j]:
            arr[k] = L[i]
            i+=1
        else:
            arr[k] = R[j]
            j+=1
        k+=1
    # Checking if any element was left in both the subarrays
    while i < n1:
        arr[k] = L[i]
        i += 1
        k += 1
 
    while j < n2:
        arr[k] = R[j]
        j += 1
        k += 1

def merge_sort(arr,lb,ub):
    if(lb < ub):
        mid = lb + (ub-lb)//2
        #same as (ub+lb)/2 but avoids overflow for large values
        #print("-->",lb,ub,mid) 
        merge_sort(arr,lb,mid) #left half
        merge_sort(arr,mid+1,ub) #right half
        '''repeated calling of merge_sort till atomic values are reached i.e.
        individual elements'''

        merge(arr,lb,ub,mid)
  

if __name__ == '__main__':
    lst = [int(i) for i in raw_input("Enter elements seperated by space :").split()]
    merge_sort(lst,0,len(lst)-1)
    print"After merge sort new list is:-\n",lst
